### ISSUE TITLE
Issues from Accessibility audit

### ISSUE THREAD

**Issue by Simon-Laux** [deltachat/deltachat-desktop#4743]
We recently got an accessibility quick scan from HAN. They took some time to discuss/test the app with us in a call and gave us a report. treefit and wofwca also made notes during the call, this issues combines points from our notes and the report they gave us. The order is roughly chronological based on the call. (might make sense to reorder and group it into areas of concern, later).

- [x] Hierarchy predictability issue. So user can move logically through the chats. They say we have 3 sections. "accounts", "chats" "messages". From search we need to focus the list of chats
	- [2.4.3 Focus Order](https://www.w3.org/WAI/WCAG22/Understanding/focus-order)
	- Visual sections in the top bar (https://github.com/deltachat/deltachat-desktop/pull/4672)
	- 2025-05-04: Further improvements, turn the sections into landmarks: https://github.com/deltachat/deltachat-desktop/pull/5067

- [x] unread badge says "1", should say "1 new message" (https://github.com/deltachat/deltachat-desktop/pull/4655)

- [x] Text in text avatar should not be read
	- https://github.com/deltachat/deltachat-desktop/pull/4661

- [x] Chat nav header (button that opens chat profile) should announce a bit more what it is (like "chat/group with name" instead of just chatname)??
	- 2025-08-04 https://github.com/deltachat/deltachat-desktop/pull/5067 should address this. It will announce "Chat with Bob section: view profile button"

- [ ] QR code button should have a link purpose
	- (announce what it does/what it is for, that it opens a dialog)
	- [2.4.4 link purpose in the standard](https://www.w3.org/WAI/WCAG22/Understanding/link-purpose-in-context)
	- Should say what you'd do with the QR code. The link purpose should be clear to the user.

- [ ] QR Code should have a useful alt/label that is not just the filename
	- [x] QR invite https://github.com/deltachat/deltachat-desktop/pull/5038
	- [x] Own secure join group qr code (qr code button)
	- [x] group join code
	- [ ] add second device code

- [ ] QR dialog: "scan to chat with" should be the header probably

- [x] Gallery: Tabs are not selectable via keyboard (https://github.com/deltachat/deltachat-desktop/pull/4675)
	They use "roving tabindex", which was the case when the audit was performed. Maybe they expected "Tab" to work there. Should we remove roving tabindex?

- [x] Gallery not sure how to escape it. -> wofwca made a tab
   Edit: also see https://github.com/deltachat/deltachat-desktop/issues/5074

- [ ] html make use of the `<time>` tag
	A `<time>`-element is missing for timestamps, which can cause inconsistent (time-based) announcements. may apply to chatlist, bubbles, daymarkers

- [x] padlock and message state - EDIT: meanwhile we do not show encrypted state at all as it is the default. instead, we show an email-icon, that is also labeled as such
	- [x] group message metadata so it is clear that it is not part of the message
	- [x] don't say padlock, instead say encrypted.
	- [x] bigger discussion: think about whether we can omit the padlock (there were discussions also with @r10s whether we should rather show broken lock on unencryped messages)
	   2025-04-21 https://github.com/deltachat/deltachat-desktop/pull/4659 removed the padlock for screen-readers
           2025-07-28 We removed the padlock, and instead show "email" for unencrypted groups, but it's `aria-hidden`. It's not clear that the message unencrypted, but apparently we can't change this only in desktop, without Android or iOS: https://github.com/deltachat/deltachat-desktop/pull/5302#discussion_r2217275888.
	- [x] current confusions
		- [x] The "lock" icon on messages. But screen reader should not read the lock.  it gives an impression that you can "unlock" the message, like "open it or something".  Don't announce it unless it's not encrypted?
		- [x] adb's mum thought it was a handbag 👜 

- [ ] message metadata contrast is too low in light theme for outgoing messages 
	- Green text on the green background, contrast is too low.

- [x] context menus should have better accessibility names
	- the "dialog group" is the name of the context menu.
	- 2025-07-29 https://github.com/deltachat/deltachat-desktop/pull/5355 should improve this, along with https://github.com/deltachat/deltachat-desktop/pull/5347 and https://github.com/deltachat/deltachat-desktop/pull/5354

- [ ] can't go back from help; close help window with escape key

- [ ] footers in dialogs should be part of form / second fieldset - their example was the disappearing messages dialog
	- disappearing messages" dialog says "footer". "save" should be in the form. The buttons should be in a fieldset probably.

- [ ] Search - voiceover does not tell that there are results
	- When you enter something to search, it should read out the results. 
	- "status message", 4.1.3. Need to implement aria to make this work. ARIA.

- [ ] message list: Reconsider ordered list.
	- Is it a list? not really a list, but

- [x] focus outline/border is not visible enough
	- we need a simple more clear focus outline, maybe look for discord for inspiration, they have a thick blue outline
	- messages
	- chatlist items (especially on selected chats it is not clearly visible)

- [ ] Profile hover information is not accessible via keyboard / screen reader (maybe include information via aria attribute into profile?)

- [x] Reactions and Context menu hover buttons on messages are not keyboard accessible
	- also their contrast is way to low to the default background. ("The grey text (#939090) on a light beige (#E6DCD4) background.")
	- note: hover context on messages should hide. Orrr, instead, show it on tab.
	- 2025-08-04: however, the actions of these buttons are already otherwise accessible, i.e. in the right-click menu of a message. Maybe this is enough?
	- EDIT: meanwhile, also arrow key works there. and for the contrast, there is the yellow border for keyboard navigation now

- [x] add keyboard shortcuts to change between zoom levels
	-  [1.4.4 Resize Text](https://www.w3.org/TR/WCAG22/#resize-text)
	- also related is to add another option for font size over `rem` that is more flexible than zoom levels: https://github.com/deltachat/deltachat-desktop/issues/890

- [x] new message is not announced
	- #4692 
	- other programs use a statusbar we could have a hidden status bar that we focus on anouncement
	- But when you are in some chat you might not want to be distracted by a busy group? so thats why it is unclear.
	- When you receive a message with window focused, it doesn't announce that you got a new message.  
		- It should receive focus, but it doesn't have to be visible.  
		- Aria could be implemented, but simply adding an element and giving it focus should work.
	- 2025-04-23 We had a chat in "Delta Chat Dev" and decided that it'd be a good start to add just a "plop" sound for new messages of the current chat, same as on mobile. This way the user is at least notified of a new message, then they can move focus to it to make the screen reader read it.
           Edit: https://github.com/deltachat/deltachat-desktop/pull/5143.
	   However, for non-current chat of the current account we still need something else: we don't show notifications for the current account's new messages when the window is in focus.
	   2025-05-23 https://github.com/deltachat/deltachat-desktop/pull/5146 will fix this.

- [x] In the chat, when you get a message, it's also not announced  
	- Maybe focusing it will work.
	- but could be annoying if you are reading other messages or writing a message
	- 2025-08-04 see previous point. The messages in the current chat are "announced" with a sound effect, and messages in the non-current chat are announced with a system notification.

- [x] encryption info is not interesting in guaranteed e2e encrypted chats, because there are info messages about the chat's state?
	- [x] still if encryption was lost, then you it becomes interessting again, as you don't always want to read the entire chat history looking for the info messages just to find out if your message was encrypted
	- 2025-07-28 We removed the padlock. See above about padlock.

### comments not actionable right now

- Dialogs, close button should be inside of form? legend fieldset
	- The close button is outside the "form" tag. Use fieldset and legend.  
	- The QR scan should be a form, even if it's not  
	- Wait now they said we should make it not a form.

- What is the situation with context menus how can you trigger them / announce them without mouse?
	- We should have a separate button for context menus? Shift + F10 is not expected by users.
	- how do other apps do this
	- what about menus in other places that don't have the hover buttons? profiles, chatlistitems, images in fullscreen media view, items in gallery, contact list and so on
	- Edit 2025-07-06 for the "Tabs" pattern, [according to W3.org](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/), it's expected that tabs can have a context menu, and that it's activated with Shift + F10. And that in this case tabs should have `aria-haspopup="menu"`:
	  > If a tab element has a popup menu, it has the property [aria-haspopup](https://w3c.github.io/aria/#aria-haspopup) set to either menu or true.
	- 2025-07-27 https://github.com/deltachat/deltachat-desktop/pull/5345

- Screenreader does not announce that it is successfully sent (aria implementation?)
	- it is a bit unclear as messages that are sent are pending at first then can change their state to failed of sent. even after they are sent they can still change to failed (if there is a could not deliver notification)
	- 2025-04-22 https://github.com/deltachat/deltachat-desktop/pull/5021 should address this, somewhat at least.
	- Also we could add the "plop" sound effect on sending a message, as we did for incoming messages (both are present on DC Android already).

- But still, the QR should maybe more somewhere else because it related to your own profile
	- that it is more clear that it is about the profile and not the chat.

- "Bypass block"-options are missing that could allow users to navigate the application more efficiently using a keyboard or screen reader. Please consider implementing “skiplinks” in strategic locations. [2.4.1 Bypass blocks](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks)
	- They suggested skip links. instead of arrow keys????

- Ensure assistive technology notifies users about status changes, particularly for recently changed elements that are not currently in focus. https://www.w3.org/WAI/WCAG22/Understanding/status-messages
- Message list arrow keys is alright.

- To further enhance the application, please review the use of semantic HTML elements. For example, an ordered list is used to structure chats inside the chat pane, which might be suboptimal. Additional suggestions were made during the accessibility audit regarding forms. Please consider using fieldsets and legend tags more consistently throughout. Remove footer tags at the bottom of forms and instead use an additional fieldset element to improve semantics.

- The chat menus in the Contact list pane and the Top bar have some inconsistencies in their menu items, despite appearing to serve the same function. The menu in the Contact pane menu has items like Pin chat, Archive chat, Mute notifications etc. The menu in the Top bar has items like Search in Chat, disappearing Messages, Mute notifications etc
  2025-07-29: see https://github.com/deltachat/deltachat-desktop/issues/5325

- on first open offer to switch to accessible high contrast theme
	- maybe offer as checkbox on the first screen?
	- high contrast theme does not exist yet
	- Maybe pick the most accessible theme by default?

- announce that help window is a new window that will be opened on links to help; the help window should stay a separate window and should not become a popup

### useful links:

https://www.w3.org/TR/WCAG22
https://webaim.org/resources/contrastchecker/

**Comment by r10s** [comment]
ftr, scan was done by https://ngi.aimsites.nl/quickscan/

**Comment by pvagner** [comment]
> * what about menus in other places that don't have the hover buttons? profiles, chatlistitems, images in fullscreen media view, items in gallery, contact list and so on

There is [aria-haspopup](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-haspopup) that allows developers and content creators to help screen reader users to identify shift+F10 context sensitive menus and other popups. So when aria-haspopup="menu" is set on a button, on a tab control, on a list item I think it's pretty natural experience to do a thing when pressing the enter key or the space key and popup the menu when pressing F10 key.
Moderate and power screen reader users are often proficient to use keyboard commands of the operating system and their screen reader, so if the popup menu is already there and is keyboard accessible, consider adding aria-haspopup where it makes sense.
From the keyboard using shift+F10 and / or the applications key on the right hand side of the keyboard is a common way on how to access such a popup menu.
With touch screen interaction press and hold gesture is commonly used to access such a popup menu.



> "Bypass block"-options are missing that could allow users to navigate the application more efficiently using a keyboard or screen reader. Please consider implementing “skiplinks” in strategic locations. [2.4.1 Bypass blocks](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks)
> 
>     * They suggested skip links. instead of arrow keys????

I'd recommend sticking with logical keyboard navigation i.e. navigating with tab and shift+tab the way you are doing now and keyboard shortcuts e.g. ctrl+m for focusing message input. Skip links are better suited for web portals where the placement of main content may not always be predictable.

Also with that comes another recommendation consider not to move focus as a result of chat state and / or notifications. For example I am chatting in a busy group conversation, I'm trying to type in my message but other group members are exchanging messages and each new message steals my focus from the chat input... That would be complicated.
Also there are desktop notifications and on most platforms screen readers have feature to read these, so I think you don't have to think of making these accessible in other way.
Perhaps it might be doable to set aria-live="polite" on the list where new messages are being appended, however it needs to be well thought and tested as every text insertion contributes to aria-live anouncements. The unpleasant side effect of such an implementation might be that when relative timestamps of existing messages change while I'm starring on a chat log, my screen reader might end up announcing times as these will get rerendered.

Then perhaps consider adding a keyboard shortcuts that will move focus to the chat log similar how ctrl+m focuses chat input. Or perhaps the fact shift+tab moves from chat input to the chat log is enough.


Thanks again for such dedication to screen reader accessibility. If you think it might be usefull to you feel free to get in touch and I'd be happy to try thinking on these and other points you might have from my perspective with you.

Greetings

Peter

**Comment by WofWca** [comment]
Thank you so much @pvagner, this is really valuable!

**Comment by pvagner** [comment]
Ah thanks.
Since last time, I have found out you have already experienced live regions within #4692

**Comment by WofWca** [comment]
Regarding the second audit: I contacted HAN Accessibility Lab again and they said it's OK for them to performed after the Tauri project has ended.

**Comment by nicodh** [comment]
Label "bug" should be more specific. This is just a collection of all issues that came up during the audit.

**Comment by nicodh** [comment]
Closing for now - most stuff is addressed and the issue is probably not useful any more. From now on let's focus on accessibility issues reported by real users.
